laboratoires/lab02-oop1/consignes/lab02.md

```diff
@@ -0,0 +1,720 @@
+
+```cpp
+#ifndef TACHE_H
+#define TACHE_H
+
+#include <string>
+
+class Tache {
+public:
+    std::string description;
+    bool fait;
+
+    Tache();
+    Tache(std::string description, bool fait);
+};
+
+#endif
+```
+
+Un constructeur **doit** toujours avoir le même nom que la classe. Il s'agit d'une méthode spéciale qui est appelée lors de l'instanciation d'un objet de notre classe. Dans notre cas, nous avons deux constructeurs : un constructeur par défaut (sans paramètre) et un constructeur qui nous permet d'initialiser tous les attributs de notre instance. Toutefois, dans notre fichier d'en-tête `.h`, nous avons seulement ajouté les prototypes des constructeurs. Il nous reste maintenant à les implémenter (c'est-à-dire, écrire le code de nos constructeurs) pour initialiser notre objet. 
+
+Créez un fichier `tache.cpp` avec le code suivant :
+
+```cpp
+#include "tache.h"
+
+Tache::Tache() {
+    this->description = "";
+    this->fait = false;
+}
+
+Tache::Tache(std::string description, bool fait) {
+    this->description = description;
+    this->fait = fait;
+}
+
+```
+
+> Concept : `this` - Comme nous l'avons déjà dit, la classe est une description : si je n'instancie pas d'objet de cette classe dans mon code, ce code ne sera jamais exécuté. Mais si j'instancie deux instances, comment faire la distinction entre les deux ? Voici un exemple :
+> ```cpp
+> int main () {
+>     std::locale::global(std::locale{ "" });
+> 
+>     Tache premier("Comprendre les classes", false);
+>     Tache seconde("Comprendre les constructeurs", false);
+> 
+>     return 0;
+> }
+> ```
+> En plus d'initialiser notre tâche d'un coup, il faut comprendre que chacun des objets (`premier` et `seconde`) sont dans deux espaces mémoires distinctes, et qu'ils ont leurs propres instance de `description` et `fait`. Toutefois, on ne peut pas connaitre d'avance le nom de toute les instances de notre classes. C++ nous offre donc `this`, qui est un pointeur vers l'instance en cours. Donc, quand je suis dans le constructeur de `seconde`, `this` pointe à la même case mémoire que `seconde` et peut donc accéder au attributs `description` et `fait` de celui-ci !
+> Pour vous convaincre que `this` est un pointeur, on peut réécrire notre constructeur de cette façon :
+> ```cpp
+> #include "tache.h"
+> 
+> Tache::Tache(std::string description, bool fait) {
+>     *this.description = description;
+>     *this.fait = fait;
+> }
+> 
+> ```
+> Ici, on utilise le déréférencement (`*this`) et on obtient donc un objet de l'instance `Tache`. On peut donc utiliser le `.` au lieu de `->`.
+
+Voici un *schéma de classe* qui représente notre `Tache` :
+
+```plantuml
+@startuml
+class Tache {
+    std::string description
+    bool fait
+
+    Tache();
+    Tache(std::string description, bool fait)
+}
+@enduml
+```
```

Comment:
J'ajouterais quelque chose pour expliquer que si on ne met pas `this`, et que notre attribut n'a pas le même nom qu'une variable locale (comme un paramètre du constructeur), ça reviendra au même. Surtout que c'est ce qu'on fait dans les exemples des slides, et possiblement dans le manuel (vu que l'auteur utilise le préfixe `_m`).